### ISSUE TITLE
docker.exe broken xcopy in same cmd window

### ISSUE THREAD

**Issue by cyuz** [docker/docker#24880]
After run docker.exe in cmd, xcopy won't give any response.

This seems related to this issues https://social.msdn.microsoft.com/Forums/vstudio/en-US/ab3c0cc7-83c2-4a86-9188-40588b7d1a52/processstart-of-xcopy-only-works-under-the-debugger?forum=netfxbcl

My current workaround is 
start /wait docker blah....

But this will make the output in another window and close

**Comment by thaJeztah** [comment]
Can you provide the exact steps to reproduce; it's currently not clear what you're trying to do. Also, please provide the output of `docker version` and `docker info`

**Comment by cyuz** [comment]
```
**docker version:**
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:20:08 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:20:08 2016
 OS/Arch:      linux/amd64

**docker info:**
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 44
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 31
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge
Kernel Version: 4.4.12-boot2docker
Operating System: Boot2Docker 1.11.2 (TCL 7.1); HEAD : a6645c3 - Wed Jun  1 22:5
9:51 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: default
ID: FRP3:74IV:BYO6:FJPR:GVJC:GUH6:OEY5:ULU5:TXSJ:WMNL:WFBI:MKS6
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug mode (client): false
Debug mode (server): true
 File Descriptors: 13
 Goroutines: 29
 System Time: 2016-07-21T12:42:26.688667697Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox

OS:
Win10 Profession x64
```

What I want to do is do a xcopy after run docker.exe
The followings display the situation

```
D:\test>dir
 磁碟區 D 中的磁碟沒有標籤。
 磁碟區序號:  58AF-8A96

 D:\test 的目錄

2016/07/21  下午 08:36    <DIR>          .
2016/07/21  下午 08:36    <DIR>          ..
2016/07/21  下午 08:35    <DIR>          test1
               0 個檔案               0 位元組
               3 個目錄  103,220,482,048 位元組可用

D:\test>xcopy test1 test2 /SYRi
test1\test.txt
已複製 1 個檔案

D:\test>dir
 磁碟區 D 中的磁碟沒有標籤。
 磁碟區序號:  58AF-8A96

 D:\test 的目錄

2016/07/21  下午 08:36    <DIR>          .
2016/07/21  下午 08:36    <DIR>          ..
2016/07/21  下午 08:35    <DIR>          test1
2016/07/21  下午 08:36    <DIR>          test2
               0 個檔案               0 位元組
               4 個目錄  103,220,482,048 位元組可用

D:\test>@FOR /f "tokens=*" %i IN ('docker-machine env') DO @%i

D:\test>docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED
STATUS              PORTS               NAMES

D:\test>xcopy test1 test3 /SYRi

D:\test>dir
 磁碟區 D 中的磁碟沒有標籤。
 磁碟區序號:  58AF-8A96

 D:\test 的目錄

2016/07/21  下午 08:36    <DIR>          .
2016/07/21  下午 08:36    <DIR>          ..
2016/07/21  下午 08:35    <DIR>          test1
2016/07/21  下午 08:36    <DIR>          test2
               0 個檔案               0 位元組
               4 個目錄  103,220,482,048 位元組可用

D:\test>
```

The second xcopy after docker.exe, didn't do any thing or give any response

**Comment by thaJeztah** [comment]
Strange, I have no idea how running `docker ps` would affect this (it only prints the output, then exits)

**Comment by thaJeztah** [comment]
Does it work if you run `xcopy` *before* `docker ps`, but *after* `@FOR /f "tokens=*" %i IN ('docker-machine env') DO @%i` ?

**Comment by cyuz** [comment]
xcopy worked find just after `@FOR /f "tokens=*" %i IN ('docker-machine env') DO @%i`.
The only problem is docker.exe.
Even call docker.exe without any parameter (just print out the help menu), xcopy still be broekn.

**Comment by thaJeztah** [comment]
@jstarks @friism @nathanleclaire any ideas?

**Comment by jstarks** [comment]
Very strange. We will try to reproduce this.

cc @darrenstahlmsft

**Comment by thaJeztah** [comment]
Note that this is on docker toolbox, but perhaps you had some insight :-)

**Comment by cyuz** [comment]
I install dockertoolbox in another win10 x64. Still meet this situations.
The system PATH
`C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\Git\cmd;;C:\Program Files\Docker Toolbox`

**Comment by nathanleclaire** [comment]
Strange bug!  Thanks for reporting.  Just to make sure I understand this bug correctly:

- If you start a new terminal and try `xcopy` it works correctly
- Then, if you run any `docker` command, even `docker --help`, `xcopy` stops working correctly

?

If you do not ever run the referenced Machine command is the bug still duplicatable in a fresh terminal session?

Restarting the terminal fixes the issue (until another `docker` command is invoked)?

This is `cmd.exe` right?  Can you duplicate this same result in Powershell?  How about in Git for Windows or Cygwin family of shells?

BTW, what's the `/SYRi`?  I am not familiar with `xcopy` so bear with me.

Thanks.

**Comment by cyuz** [comment]
1. xcopy will worked fine at another new **cmd** terminal
2. xcopy will be broken again after docker

about xcopy parameters:
- /s : Copies directories and subdirectories, unless they are empty. If you omit /s, xcopy works within a single directory.
- /y : Suppresses prompting to confirm that you want to overwrite an existing destination file.
- /r : Copies read-only files.
- /i : If Source is a directory or contains wildcards and Destination does not exist, xcopy assumes destination specifies a directory name and creates a new directory. Then, xcopy copies all specified files into the new directory. By default, xcopy prompts you to specify whether Destination is a file or a directory.
These parameters equal to copy a directory to another directory without any prompt.


cmd terminal : observe this bug
powershell : also observe this bug 
git bash : **no this bug**
cygwin terminal : **no this bug**

**Comment by darrenstahlmsft** [comment]
I've got a local repro, and am investigating.

**Comment by darrenstahlmsft** [comment]
I've tracked this down to a Windows bug. SetConsoleMode remembers invalid bits even on a failure. This causes all applications that rely on setting the console mode to fail after docker.exe is run.

This affects all docker versions after 1.11.1 where virtual terminal input was added. I'm not sure on the process for backporting fixes, but think that this should likely be patched in 1.11. Maintainers, WDYT?

 PR with fix is submitted.
